File: src/main/java/pl/dmichalski/algorithms/_4_sum_zero/SumToZeroService2.java
package pl.dmichalski.algorithms._4_sum_zero;

class SumToZeroService2 {

    /**
     * O(n) complexity
     */
    int[] sumZero(int[] values) {
        int left = 0;
        int right = values.length - 1;

        while (left < right) {
            int sum = values[left] + values[right];
            if (sum == 0) {
                return new int[]{values[left], values[right]};
            } else if (sum > 0) {
                right--;
            } else {
                left++;
            }
        }

        return null;
    }

}

File: src/main/kotlin/pl/dmichalski/algorithms/_4_sum_zero/SumToZeroService2.kt
package pl.dmichalski.algorithms._4_sum_zero

internal class SumToZeroService2 {

    /**
     * O(n) time complexity
     */
    fun sumZero(values: IntArray): IntArray? {
        var left = 0
        var right = values.size - 1
        while (left < right) {
            val sum = values[left] + values[right]
            when {
                sum == 0 -> {
                    return intArrayOf(values[left], values[right])
                }
                sum > 0 -> {
                    right--
                }
                else -> {
                    left++
                }
            }
        }
        return null
    }

}
